core/server/src/main/java/alluxio/master/file/FileSystemMaster.java

```diff
@@ -2814,17 +2814,41 @@ public FileSystemCommand workerHeartbeat(long workerId, List<Long> persistedFile
       inode.setLastModificationTimeMs(opTimeMs);
     }
     if (options.getTtl() != null) {
-      Preconditions.checkArgument(inode.isFile(), PreconditionMessage.TTL_ONLY_FOR_FILE);
       long ttl = options.getTtl();
-      InodeFile file = (InodeFile) inode;
-      if (file.getTtl() != ttl) {
-        mTtlBuckets.remove(file);
-        file.setTtl(ttl);
-        mTtlBuckets.insert(file);
-        file.setLastModificationTimeMs(opTimeMs);
-        file.setTtlAction(options.getTtlAction());
+      if (inode.isDirectory()) {
+        InodeDirectory directory = (InodeDirectory) inode;
+        InodeDirectory parent = inodePath.getParentInodeDirectory();
+        if (parent != null) {
+          System.out.println("parent:" + parent.getTtl() + " " + ttl);
+          Preconditions.checkArgument(parent.getTtl() < 0 || ttl <= parent.getTtl(),
+              PreconditionMessage.TTL_FILE_SHOULD_LESS_THAN_PARENT_DIRECTORY);
+        }
+        if (directory.getTtl() != ttl) {
+          mTtlBuckets.remove(directory);
+          directory.setTtl(ttl);
+          mTtlBuckets.insert(directory);
+          directory.setLastModificationTimeMs(opTimeMs);
+          directory.setTtlAction(options.getTtlAction());
+        }
+        Set<Inode<?>> children = directory.getChildren();
+        for (Inode<?> child : children) {
+          AlluxioURI alluxioURI = new AlluxioURI(inodePath.getUri().getPath()
+              + "/" + child.getName());
+          setAttribute(alluxioURI, options);
```

Comment:
If set the value of mRecursive to true, in method setAttributeAndJournal the parent inode should first run setAttributeInternal. Because I think the file's ttl should less than parent directoryInode's ttl, So, I add  Preconditions.checkArgument(parent.getTtl() < 0 || ttl <= parent.getTtl(),
          PreconditionMessage.TTL_FILE_SHOULD_LESS_THAN_PARENT_DIRECTORY) to guarantee it.